tensorflow_probability/python/distributions/empirical.py

```diff
@@ -0,0 +1,186 @@
+        performance. When `False` invalid inputs may silently render incorrect
+        outputs.
+      allow_nan_stats: Python `bool`, default `True`. When `True`, statistics
+        (e.g., mean, mode, variance) use the value "`NaN`" to indicate the
+        result is undefined. When `False`, an exception is raised if one or
+        more of the statistic's batch members are undefined.
+      name: Python `str` name prefixed to Ops created by this class.
+    """
+    parameters = locals()
+    with tf.name_scope(name, values=[samples]):
+      self._samples = tf.convert_to_tensor(samples, name="samples")
+      samples_shape_static = self._samples.shape.with_rank_at_least(1)
+      self._num_sample = tf.shape(self._samples)[0]
+
+    super(Empirical, self).__init__(
+        dtype=self._samples.dtype,
+        reparameterization_type=reparameterization.FULLY_REPARAMETERIZED,
+        validate_args=validate_args,
+        allow_nan_stats=allow_nan_stats,
+        parameters=parameters,
+        graph_parents=[self._samples, self._num_sample],
+        name=name)
+
+  @staticmethod
+  def _param_shapes(sample_shape):
+    return {"samples": tf.convert_to_tensor(sample_shape, dtype=tf.int32)}
+
+  @property
+  def samples(self):
+    """Distribution parameter."""
+    return self._samples
+
+  @property
+  def num_sample(self):
+    """Number of samples."""
+    return self._num_sample
+
+  def _batch_shape_tensor(self):
+    return tf.constant([], dtype=tf.int32)
+
+  def _batch_shape(self):
+    return tf.TensorShape([])
+
+  def _event_shape_tensor(self):
+    return tf.shape(self.samples)[1:]
+
+  def _event_shape(self):
+    return self.samples.shape[1:]
+
+  def _mean(self):
+    return tf.reduce_mean(self.samples, 0)
+
+  def _mode(self):
+    y, idx = gen_array_ops.unique_v2(self.samples, axis=[0])
+    unique_count = tf.unique_with_counts(idx).count
+    return tf.gather(y, tf.argmax(unique_count))
+
+  def _stddev(self):
+    # broadcasting n x shape - shape = n x shape
+    r = self.samples - self.mean()
+    var = tf.cast(tf.reduce_mean(tf.square(r), 0), tf.float32)
+    return tf.sqrt(var)
+
+  def _sample_n(self, n, seed=None):
+    cat = Categorical(logits=tf.zeros([self.num_sample]))
+    indices = cat.sample(n, seed)
+    tensor = tf.gather(self.samples, indices)
+    return tensor
+
+  def _log_prob(self, event):
+    event = tf.convert_to_tensor(event, name='event', dtype=self.dtype)
+    if self.validate_args:
+      event = util.embed_check_integer_casting_closed(
+        event, target_dtype=tf.bool)
+
+    # Raise an error when the labels shape does not equal the event shape.
+    # Note: This could break users who call this with bad labels, but
+    # disregard the bad results.
+    labels_static_shape = event.get_shape()
+    static_shapes_fully_defined = (
+            labels_static_shape.is_fully_defined() and
+            self.samples.get_shape()[1:].is_fully_defined())
```

Comment:
prefer `self.event_shape` to `self.samples.get_shape()[1:]`, and `self.event_shape.ndims` to `self.samples.get_shape().ndims - 1`.

Reply:
fixed.